unity/JetBrains.Rider.Unity.Editor/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs

```diff
@@ -332,54 +333,89 @@ private static void ApplyCustomReference(string name, XElement projectContentEle
+                else if (foundVersion == version)
+                  ourLogger.Verbose("Found TargetFrameworkVersion {0} equals the one set-by-Unity itself {1}",
+                    foundVersion, version);
+                else if (ourLogger.IsVersboseEnabled())
+                {
+                  var message = $"Rider may require \".NET Framework {version} Developer Pack\", which is not installed.";
+                  Debug.Log(message);
+                }
               }
             }
           }
-          targetFrameworkVersion.SetValue("v" + version);
-        }
-        else
-        {
-          var version = PluginSettings.TargetFrameworkVersionOldMono;
-          if (!PluginSettings.OverrideTargetFrameworkVersionOldMono)
+          catch (Exception e)
           {
-            if (PluginSettings.SystemInfoRiderPlugin.operatingSystemFamily == OperatingSystemFamilyRider.Windows)
+            ourLogger.Log(LoggingLevel.WARN, "Fail to FixTargetFrameworkVersion", e);
+          }
+
+          if (UnityUtils.ScriptingRuntime > 0)
+          {
+            if (PluginSettings.OverrideTargetFrameworkVersion)
             {
-              var versions = PluginSettings.GetInstalledNetFrameworks();
-              if (versions.Any())
-              {
-                version = versions.OrderBy(v => new Version(v)).First();
-              }
+              return "v" + PluginSettings.TargetFrameworkVersion;
             }
           }
-          targetFrameworkVersion.SetValue("v" + version);
+          else
+          {
+            if (PluginSettings.OverrideTargetFrameworkVersionOldMono)
+            {
+              return "v" + PluginSettings.TargetFrameworkVersionOldMono;;
+            }
+          }
+
+          return "v" + version;
         }
+      );
     }
 
     private static void SetLangVersion(XElement projectElement, XNamespace xmlns)
     {
       // Set the C# language level, so Rider doesn't have to guess (although it does a good job)
       // VSTU sets this, and I think newer versions of Unity do too (should check which version)
-      SetOrUpdateProperty(projectElement, xmlns, "LangVersion", GetLanguageLevel());
+      SetOrUpdateProperty(projectElement, xmlns, "LangVersion", existing =>
+      {
+        var expected = GetExpectedLanguageLevel();
+        if (expected == "latest" || existing == "latest")
+          return "latest";
+
+        // Only use our version if it's not already set, or it's less than what we would set
+        // Note that if existing is "default", we'll override it
+        var currentIsParsed = decimal.TryParse(existing, NumberStyles.Number, CultureInfo.InvariantCulture, out var currentLanguageLevel);
+        var expectedIsParsed = decimal.TryParse(expected,NumberStyles.Number, CultureInfo.InvariantCulture, out var expectedLanguageLevel);
+        if (currentIsParsed && expectedIsParsed && currentLanguageLevel < expectedLanguageLevel
+            || !currentIsParsed
+            )
```

Comment:
It might be better to use `float` here instead of `decimal`. And perhaps check the return of `TryParse` and set defaults, to make the conditional easier? If you can't parse current, set it to `float.MinValue`, if you can't parse expected, set it to `float.MaxValue`, then it's a simple comparison for all conditions?

Reply:
when debugging floats look weird 3.5 becomes 3.499991 ))

Reply:
Yikes! Is that a bug in the debugger or mono?